src/db.rs

```diff
@@ -42,295 +44,114 @@ pub struct IdMap {
-                self.write_path(&graph.file(fileid).name)?;
+                let entry = DbEntry::File(graph.file(fileid).name.to_owned());
+                serde_cbor::ser::to_writer(&mut self.w, &entry)?;
                 id
             }
         };
         Ok(id)
     }
 
-    pub fn write_build(&mut self, graph: &Graph, id: BuildId, hash: Hash) -> std::io::Result<()> {
+    pub fn write_build(&mut self, graph: &Graph, id: BuildId, hash: Hash) -> anyhow::Result<()> {
         let build = graph.build(id);
-        let mut buf = WriteBuf::new();
-        let outs = build.outs();
-        let mark = (outs.len() as u16) | 0b1000_0000_0000_0000;
-        buf.write_u16(mark);
-        for &out in outs {
-            let id = self.ensure_id(graph, out)?;
-            buf.write_id(id);
-        }
-
-        let deps = build.discovered_ins();
-        buf.write_u16(deps.len() as u16);
-        for &dep in deps {
-            let id = self.ensure_id(graph, dep)?;
-            buf.write_id(id);
-        }
-
-        buf.write_u64(hash.0);
-
-        buf.flush(&mut self.w)
-    }
-}
-
-struct Reader<'a> {
-    r: BufReader<&'a mut File>,
-    ids: IdMap,
-    graph: &'a mut Graph,
-    hashes: &'a mut Hashes,
-}
-
-impl<'a> Reader<'a> {
-    fn read_u16(&mut self) -> std::io::Result<u16> {
-        let mut buf: [u8; 2] = [0; 2];
-        self.r.read_exact(&mut buf[..])?;
-        Ok(u16::from_le_bytes(buf))
-    }
-
-    fn read_u24(&mut self) -> std::io::Result<u32> {
-        let mut buf: [u8; 4] = [0; 4];
-        self.r.read_exact(&mut buf[..3])?;
-        Ok(u32::from_le_bytes(buf))
-    }
-
-    fn read_u64(&mut self) -> std::io::Result<u64> {
-        let mut buf: [u8; 8] = [0; 8];
-        self.r.read_exact(&mut buf)?;
-        Ok(u64::from_le_bytes(buf))
-    }
-
-    fn read_id(&mut self) -> std::io::Result<Id> {
-        self.read_u24().map(Id)
-    }
-
-    fn read_str(&mut self, len: usize) -> std::io::Result<String> {
-        let mut buf = Vec::with_capacity(len);
-        // Safety: buf contents are uninitialized here, but we never read them
-        // before initialization.
-        // TODO: clippy says this is still UB, yuck.
-        unsafe { buf.set_len(len) };
-        self.r.read_exact(buf.as_mut_slice())?;
-        Ok(unsafe { String::from_utf8_unchecked(buf) })
-    }
-
-    fn read_path(&mut self, len: usize) -> std::io::Result<()> {
-        let mut name = self.read_str(len)?;
-        let fileid = self.graph.file_id(&mut name);
-        let dbid = self.ids.fileids.push(fileid);
-        self.ids.db_ids.insert(fileid, dbid);
+        let outs = build
+            .outs()
+            .iter()
+            .map(|&file_id| self.ensure_id(graph, file_id))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+        let deps = build
+            .discovered_ins()
+            .iter()
+            .map(|&file_id| self.ensure_id(graph, file_id))
+            .collect::<anyhow::Result<Vec<_>>>()?;
```

Comment:
I note that your new code allocs more (the intermediate Vecs here), but I think writing isn't on the critical path anyway.